### DISCUSSION TITLE
Instantiating models using DeclarativeBase and Mapping does not check types

### DISCUSSION THREAD

**Discussion by david-caro** [sqlalchemy/sqlalchemy#11169] [Usage Questions] [answered]
### Ensure stubs packages are not installed

- [X] No sqlalchemy stub packages is installed (both `sqlalchemy-stubs` and `sqlalchemy2-stubs` are not compatible with v2)

### Verify if the api is typed

- [X] The api is not in a module listed in [#6810](https://github.com/sqlalchemy/sqlalchemy/issues/6810) so it should pass type checking

### Describe the typing issue

Wen using mapped columns for a model, mypy does not type check the parameters passed when instantiating the model.

### To Reproduce

```python
from sqlalchemy.orm import Mapped, DeclarativeBase


class Base(DeclarativeBase):
    pass


class MyModel(Base):
    my_int: Mapped[int]
    my_str: Mapped[str]


my_instance = MyModel(idontexist="something", my_int="notanint", my_str=42)

# mypy --strict test.py
```

### Error

```
Success: no issues found in 1 source file
``` 

### Versions

- OS: Fedora 39
- Python: 3.12.2
- SQLAlchemy: 2.0.28
- Type checker (eg: mypy 0.991, pyright 1.1.290, etc): mypy 1.9.0

### Additional context

When trying to assign the specific attributes directly, the types are correct:
```
my_instance.my_int = "notanint"
```

Complains with:
```
test.py:14: error: Incompatible types in assignment (expression has type "str", variable has type "SQLCoreOperations[int] | int")  [assignment]
Found 1 error in 1 file (checked 1 source file)
```

I'm not sure if this is not supported or I'm doing something wrong (did not find it in the docs or issues).

Thanks!

**Comment by CaselIT** [comment]
Hi,

The default init does not check the types. You have the option of using dataclasses to have the init type checked as expected.

**Reply by david-caro** [reply]
Awesome! It works :+1:  for reference now I have:

```python
from sqlalchemy.orm import Mapped, DeclarativeBase
from dataclasses import dataclass


class Base(DeclarativeBase):
    pass


@dataclass
class MyModel(Base):
    my_int: Mapped[int]
    my_str: Mapped[str]


my_instance = MyModel(idontexist="something", my_int="notanint", my_str=42)  # this fails

# mypy --strict test.py
```

**Reply by david-caro** [reply]
Oh no!

But it fails also with the right params:
```
$ mypy --strict test.py 
test.py:15: error: Argument "my_int" to "MyModel" has incompatible type "int"; expected "Mapped[int]"  [arg-type]
test.py:15: error: Argument "my_str" to "MyModel" has incompatible type "str"; expected "Mapped[str]"  [arg-type]
Found 2 errors in 1 file (checked 1 source file)
```

**Reply by david-caro** [reply]
Interesting enough, vscode reveals that the signature is:
```
class MyModel(
    my_int: SQLCoreOperations[int] | int,
    my_str: SQLCoreOperations[str] | str
)
```
And that should work, but mypy complains :/

**Reply by david-caro** [reply]
Mypy says this with `reveal_type`:
```
test.py:15: note: Revealed type is "def (my_int: sqlalchemy.orm.base.Mapped[builtins.int], my_str: sqlalchemy.orm.base.Mapped[builtins.str]) -> test.MyModel"
```

**Reply by david-caro** [reply]
pyright gets it correct though, it passes when it should, and if I pass a sting to the integer parameter I get error:
```
pyright test.py 
 test.py:15:30 - error: Argument of type "Literal['somestring']" cannot be assigned to parameter "my_int" of type "SQLCoreOperations[int] | int" in function "__init__"
    Type "Literal['somestring']" cannot be assigned to type "SQLCoreOperations[int] | int"
      "Literal['somestring']" is incompatible with "SQLCoreOperations[int]"
      "Literal['somestring']" is incompatible with "int" (reportArgumentType)
1 error, 0 warnings, 0 informations 
```

That shows the type it resolves for the parameter is `SQLCoreOperations[int] | int`

**Reply by CaselIT** [reply]
You need to use mapped as dataclass.
It's extensively documented here https://docs.sqlalchemy.org/en/20/orm/dataclasses.html

**Reply by david-caro** [reply, marked as answer]
I see, it's a sqlalchemy thing, not bulit-in dataclasses xd, okok, so a working example now for me is:
```python
from sqlalchemy.orm import (
    Mapped,
    DeclarativeBase,
    MappedAsDataclass,
    relationship,
    mapped_column,
)


class Base(MappedAsDataclass, DeclarativeBase):
    pass


class Role(Base):
    __tablename__ = "roles"

    id: Mapped[int] = mapped_column(primary_key=True, init=False)
    name: Mapped[str]


class User(Base):
    __tablename__ = "users"

    id: Mapped[int] = mapped_column(primary_key=True, init=False)
    name: Mapped[str]
    role: Mapped[Role] = relationship("Role", back_populates="roles", init=False)


my_role = Role(name="somerole")
my_user = User(name="User1")
```

Thanks a lot!

**Reply by CaselIT** [reply]
These are dataclasses, but you need to use the sqlalchemy base or decorator for type checkers to like them. Otherwise you may gave unexpected type errors like you reported
